tls/s2n_async_pkey.c

```diff
@@ -279,7 +282,16 @@ S2N_RESULT s2n_async_pkey_sign_sync(struct s2n_connection *conn, s2n_signature_a
     RESULT_GUARD(s2n_pkey_size(pkey, &maximum_signature_length));
     RESULT_GUARD_POSIX(s2n_alloc(&signed_content, maximum_signature_length));
 
-    RESULT_GUARD_POSIX(s2n_pkey_sign(pkey, sig_alg, digest, &signed_content));
+    RESULT_ENSURE_REF(conn->config);
+    if (conn->config->verify_after_sign) {
+        DEFER_CLEANUP(struct s2n_hash_state digest_for_verify, s2n_hash_free);
+        RESULT_GUARD_POSIX(s2n_hash_new(&digest_for_verify));
+        RESULT_GUARD_POSIX(s2n_hash_copy(&digest_for_verify, digest));
+        RESULT_GUARD_POSIX(s2n_pkey_sign(pkey, sig_alg, digest, &signed_content));
+        RESULT_GUARD(s2n_async_pkey_verify_signature(conn, sig_alg, &digest_for_verify, &signed_content));
+    } else {
+        RESULT_GUARD_POSIX(s2n_pkey_sign(pkey, sig_alg, digest, &signed_content));
+    }
 
     RESULT_GUARD_POSIX(on_complete(conn, &signed_content));
 
```

Comment:
I'm not sure if this would be worth doing, but one way we could potentially test the failing case of verify_after_sign might be to have a function, `verify_signed_content` or something, that you always pass the signed content to. And, if verify_after_sign is false, it would no-op. So in the test, you could call sign, mess up the signed content, and call the verify function with verify_after_sign set to true and false. 

This still wouldn't really test the s2n_async_pkey_sign_sync function itself though, so I'm not sure if it makes sense to do this...

Reply:
Yeah I'm not sure that would really be worth it, since we already know that s2n_async_pkey_verify_signature works from the async version of the tests. It would basically just test the if/else condition 😅